Mtd 11.0 wave 2

ShreeOpusVL opened the pull request:
Note from PA - I have re-done 11.0 branches and the mtd-11.0 branch from a clean 8.0 base, cherry picking all commits which were included in the previous PR. This was to get git out of the messy state it was in.

Once this is merged in and we can be sure that we didn't miss anything - branches `origin/mtd-11.0-wave-1` and `origin/mtd_w1` can be deleted. As far as we are concerned with PRs - those branches can be ignored for now

ShreeOpusVL committed:
[INI][O1758]- MTD Initial commit

This commit contains successful completion of Hello world API and hello
application API. There are some changes which are still required but
at least we can establish a connection. The hello user step one getting
the authorisation code successfully achieved (still need to handle the
exceptions) and step two gaining the access code and refresh code also
works. Now need to receive the hello user message (step three from the
connection.)

ShreeOpusVL committed:
[IMP][O1758]- MTD development step 3 complete.

Getting hello user message from HMRC completed and some UI changes have
also been made.

ShreeOpusVL committed:
[IMP][O1758] Step 4 refresh development complete

user groups created added new data files for hello world. partial
logging added to hello world.

ShreeOpusVL committed:
[IMP][O1758] pr changes

some more log messages added and changes to some existing log messages

ShreeOpusVL committed:
[IMP][O1758] Missed few pr changes - now added

ShreeOpusVL committed:
[IMP][O1758] Pep8 changes

ShreeOpusVL committed:
[IMP][O1758] fixing the error on hello user

While testing found an error which is now been fixed. and made few minor
changes

ShreeOpusVL committed:
[IMP][O1758] Fall back for request tracker added

ShreeOpusVL committed:
[IMP][O1758] Changes to the lables

ShreeOpusVL committed:
[IMP][O1758] fixing typo

ShreeOpusVL committed:
[IMP][O1758]Adding accidentally removed files

ShreeOpusVL committed:
[IMP][O1758]Error fixing

ShreeOpusVL committed:
[IMP][O1758] Changes to variable name.

Changes request by Manish

ShreeOpusVL committed:
[IMP][O1758] Security changes for mtd admin

MTD admin should alsonot be able to create any endpoints.

ShreeOpusVL committed:
[IMP][O1758]security changes for mtd user

Mtd user should not be able to make changes to the HMRC configuration
and should have create access to api_request_tracker and api_tokens

ShreeOpusVL committed:
[IMP][O1758]logging message stating what happend

and why it could have happened

ShreeOpusVL committed:
[IMP][O1758]Reson for model existence added

ShreeOpusVL committed:
[IMP][O1758] Reason for model existence added

ShreeOpusVL committed:
[IMP][O1758] typo and refactoring each if to a single call

ShreeOpusVL committed:
[IMP][O1758] Variable name change

ShreeOpusVL committed:
[IMP][O1758] authorisation url construction

Making it simple to understand the construction of the authorisation url

ShreeOpusVL committed:
[IMP][O1758] Variable name change.

ShreeOpusVL committed:
[IMP][O1758]adding trailing comma

dedent the parenthases

ShreeOpusVL committed:
[IMP][O1758] Adding explanation for a dict

logger string changes for readability

ShreeOpusVL committed:
[IMP][O1758] Removed hard coded menu_id (left in by mistake)

ShreeOpusVL committed:
[IMP][O1758] adding missing return

and some logger message change

ShreeOpusVL committed:
[IMP][O1758] change as per pr

ShreeOpusVL committed:
[IMP][O1758] changed datetime.now to datetime.utcnow

and using named parameters for .format to make it easier to read

ShreeOpusVL committed:
[IMP][O1758] adding missing import and vars for format

ShreeOpusVL committed:
[IMP][O1758] Formatting date time field from database to date time object

and then comparing the object rather than formatting date time into
string for comperison. Also changed the name of a variable

PeterAlabaster commented on account_mtd/controllers/main.py:
```diff
@@ -0,0 +1,40 @@
+        
+        # to determine which API the authorization code has been received for.
+        api_tracker = http.request.env['mtd.api_request_tracker'].search([('closed', '=', False)])
+        if len(api_tracker) != 1:
```
Should this not read more like:
```
if not api_tracker:
    logger.warn("...")
    raise exceptions.Warning()
elif len(api_tracker) > 1:
    logger.info("...")
    ....
    do something
    ...
else:
    for record in api_tracker:
    ....
    do something
    ....
```
so that handling for none, one, and more than one is seperate? I'm just suggesting this based on the comment below about `user should never be in a state where they found no tracker record`

Nick-OpusVL replied:
Agreed with @PeterAlabaster .  It will keep comments, if any are necessary at all, brief.  The log message can be more focussed too.

PeterAlabaster commented on account_mtd/models/mtd_api.py:
```diff
@@ -0,0 +1,9 @@
+    _name = 'mtd.api'
+    _description = "list of api table"
+
+    name = fields.Char('Name', required=True, readonly=True)
```
My previous review still stands on all field declarations:
`name = fields.Char()`  is the same as `name = fields.Char('Name')` and `name = fields.Char(string='Name')`.
The only times where a string parameter is required in a field is if the string differs from the fieldname. 

See https://www.odoo.com/documentation/10.0/reference/orm.html#creating-models
```
By default, the field's label (user-visible name) is a capitalized version of the field name, this can be overridden with the string parameter:
```

So I don't think some of these are totally necessary

Nick-OpusVL replied:
I'm still in two minds about this.  I think @james-opusvl is in the "be explicit" camp, whereas @PeterAlabaster is saying lean on the defaults.  I can see benefits to both

I would say if you're labelling them explicitly, use the kwarg `string='Label'` rather than leaning on the positional arg order, which varies depending on the type of field.  No need to go onto a new line if all you're passing in is `string` but do so if it takes the line over 80 chars.

PeterAlabaster commented on account_mtd/models/mtd_api_request_tracker.py:
```diff
@@ -0,0 +1,27 @@
+from odoo import models, fields, api
+
+################################################################
```
I would suggest making this a class docstring like so:
```
class MtdApiRequestTracker(models.Model):
    """
    This model is created to keep track of any pending requests
    for access token. If one request is in process we can not
    send a request for a new one. This also contains information
    regarding where the request was sent from so when redirected
    back to Odoo we redirect user back to where they started from
    """
    _name = 'mtd.api_request_tracker'
    _description = "authorisation request table"
```

PeterAlabaster commented on account_mtd/models/mtd_api_tokens.py:
```diff
@@ -0,0 +1,25 @@
+from odoo import models, fields, api
+
+##############################################################
```
The same applies for this one. Also suggest changing `module` to `model`

PeterAlabaster commented on account_mtd/models/mtd_hello_world.py:
```diff
@@ -0,0 +1,456 @@
+    which_button_type_clicked = fields.Char(string="which_button")
+    path = fields.Char(string="sandbox_url")
+    current_record = fields.Char()
+    # need a variable which will keep the record id so that when we
```
This sentence cuts out half way through

PeterAlabaster commented on account_mtd/models/mtd_hello_world.py:
```diff
@@ -0,0 +1,456 @@
+            ('res_id', '=', self.id), ('model', '=', 'mtd.hello_world')
+        ])
+
+        if endpoint_record.name == "mtd_hello_world_endpoint":
```
I think there is a more pythonic way of doing this, I will sit at your desk to run through it

PeterAlabaster replied:
In fact, is endpoint_record something which a user can edit? in which case it is probably better to have it hard-coded like this.

PeterAlabaster replied:
```
mapping = {
    'mtd_hello_world_endpoint': '_handle_mtd_hello_world_endpoint',
    'mtd_hello_application_endpoint': '_handle_mtd_hello_application_endpoint',
    'mtd_hello_user_endpoint': '_handle_mtd_hello_user_endpoint'
}
if not mapping.get(endpoint_record.name):
    raise exceptions.Warning("Could not connect to HMRC! \nThis is not a valid HMRC service connection")
return getattr(self, mapping.get(endpoint_record.name))
```
I think should work

Nick-OpusVL replied:
Yep another step in the right direction (except you've not called the method returned by getattr()), but why not:

```
request_handlers = {
   ...
}
try:
    handler_name = request_handlers[endpoint_record.name]
except KeyError:
    raise exceptions.Warning(...)
else:
    handle_request = getattr(self, handler_name)
    return handle_request()
```

or

```
handler_name = request_handlers.get(endpoint_record.name)
if handler_name:
    handle_request = getattr(self, handler_name)
    return handle_request()
else:
    raise exceptions.Warning(...)
```


It's also possible to define handler_name in the above thus:

```
handler_name = "_handle_" + endpoint_record.name
try:
    handle_request = getattr(self, handler_name)
except AttributeError:
    raise exceptions.Warning()
else:
    return handle_request()
```

BUT...

I think this actually fits the strategy pattern [1] but I don't know where the split between classes (or models in Odoo's case) would go without a much better understanding of how all these bits of logic and data are interacting.

[1] https://en.wikipedia.org/wiki/Strategy_pattern

PeterAlabaster commented on account_mtd/models/mtd_hello_world.py:
```diff
@@ -0,0 +1,456 @@
+    _name = 'mtd.hello_world'
+    _description = "Hello world to test connection between Odoo and HMRC"
+    
+    _logger = logging.getLogger(__name__)
```
if you make _logger global (put it just under `from urllib.parse import urlparse`) then you can change every instance of `self._logger()` to just `_logger`

Nick-OpusVL replied:
Seconded.

Leave a couple of blank lines either side of it for pep8 though.

PeterAlabaster commented on account_mtd/models/mtd_hello_world.py:
```diff
@@ -0,0 +1,456 @@
+            else:
+                return self.get_user_authorisation()
+
+    def _json_command(self, command, timeout=3, record_id=None):
```
These are pretty hefty functions. Is  there any way that these could be reduced, like having common shared functionality (it may not be shared yet, but might be in future?) split out into smaller functions?

Nick-OpusVL replied:
Doesn't need to be shared functionality to be worthy of being in a method, but common stuff (as long as it's not only common by coincidence) should be.  If something is a sequence of things that need to be done.

```
_this()
_that()
_the_other()
```

and each of those functions' bodies explain *how* to do this, that and the other, it's much easier to read.  Just be careful not to mix levels of abstraction.

PeterAlabaster reviewed (changes requested):
Much better and cleaner this time round, though please see my comments

ShreeOpusVL committed:
[IMP][O1758] removing field declaration of string where not required

ShreeOpusVL committed:
[IMP][O1758]chanign comments to a class docstring

ShreeOpusVL committed:
[IMP][O1758] Removal of variable which is no longer in use

ShreeOpusVL committed:
[IMP][O1758] changing the _logger to a global variable

Nick-OpusVL commented on account_mtd/controllers/main.py:
```diff
@@ -0,0 +1,40 @@
+    def get_user_authorization(self, **args):
+        
+        # to determine which API the authorization code has been received for.
+        api_tracker = http.request.env['mtd.api_request_tracker'].search([('closed', '=', False)])
```
Pluralise this name - we can't assert there is only one at this stage.

Nick-OpusVL commented on account_mtd/controllers/main.py:
```diff
@@ -0,0 +1,40 @@
+                "\nWe can have more than one if there was an initial request and authorisation it was never completed."+
+                "\nIf in this state we need to reset the tracker and get user to create a new connection " +
+                "\napi_tracer = {}".format(api_tracker))
+            for record in api_tracker:
```
Equivalent:

```
api_trackers.write({'closed': True})
```

(I assume you now mean closed = True as that's what you're searching on earlier, apologies if not).

Nick-OpusVL replied:
The log messages are a bit verbose.  What you're about to do or the results of what you've just done are fine, but don't embed pseudocode in them.

Nick-OpusVL commented on account_mtd/models/mtd_api_tokens.py:
```diff
@@ -0,0 +1,25 @@
+    _name = 'mtd.api_tokens'
+    _description = "api token table"
+    
+    api_id = fields.Char('API Id', required=True)
```
How does this map onto the HMRC API?

Nick-OpusVL commented on account_mtd/models/mtd_api_tokens.py:
```diff
@@ -0,0 +1,25 @@
+    _description = "api token table"
+    
+    api_id = fields.Char('API Id', required=True)
+    api_name = fields.Char('API Name', required=True)
```
How does this map onto the HMRC API?  I presume it's different from the api_id.

Nick-OpusVL commented on account_mtd/models/mtd_api_tokens.py:
```diff
@@ -0,0 +1,25 @@
+    authorisation_code = fields.Char('Authorisation Code', required=True)
+    access_token = fields.Char('Access Token')
+    refresh_token = fields.Char('Refresh Token')
+    expires_in = fields.Char('Expires In')
```
Can you explain what `expires_in` holds?  Maybe an example or two?

ShreeOpusVL replied:
This is the time sent by HMRC to say how long the access token is valid for. 
Example response
{
  "access_token": "QGbWG8KckncuwwD4uYXgWxF4HQvuPmrmUqKgkpQP",
  "token_type": "bearer",
  "expires_in": 14400,
  "refresh_token": "unJkSs5cvs8CS9E4DLvTkNhcRBq9BwUPm23cr3pF",
  "scope": "read:employment"
}

Nick-OpusVL replied:
What unit is it in?

ShreeOpusVL replied:
In hours

Nick-OpusVL replied:
`expires_in_hours` ?

I have other thoughts about this but they're dependent on how it's used.

ShreeOpusVL replied:
At the moment the value entered is not being used. I am just storing the information gained back from HMRC

Nick-OpusVL commented on account_mtd/models/mtd_hmrc_configuration.py:
```diff
@@ -0,0 +1,23 @@
+    redirect_url = fields.Char('Redirect URL', required=True)
+    access_token = fields.Char('Access Token')
+    refresh_token = fields.Char('Refresh Token')
+    state = fields.Char('State')
```
Please add newline at end.

Nick-OpusVL commented on account_mtd/models/mtd_hello_world.py:
```diff
@@ -0,0 +1,456 @@
+    scope = fields.Char(related="api_name.scope")
+    response_from_hmrc = fields.Text(string="Response From HMRC", readonly=True)
+    which_button_type_clicked = fields.Char(string="which_button")
+    path = fields.Char(string="sandbox_url")
```
What's the purpose of this?

ShreeOpusVL replied:
All endpoints have a specific HMRC url to connect to e.g 
the HMRC url is - https://test-api.service.hmrc.gov.uk/
if in hello world endpoint the HMRC url to connect to is HMRC url + "/hello/world"
for hello application end point the HMRC url to connect to is HMRC url + "/hello/application"
for hello user end point the HMRC url to connect to is HMRC url + "/hello/user"

Nick-OpusVL commented on account_mtd/models/mtd_hello_world.py:
```diff
@@ -0,0 +1,456 @@
+    response_from_hmrc = fields.Text(string="Response From HMRC", readonly=True)
+    which_button_type_clicked = fields.Char(string="which_button")
+    path = fields.Char(string="sandbox_url")
+    current_record = fields.Char()
```
What's the purpose of this?

ShreeOpusVL replied:
I have already removed this. Don't need it any more

Nick-OpusVL commented on account_mtd/models/mtd_api_request_tracker.py:
```diff
@@ -0,0 +1,27 @@
+#################################################################
+
+class MtdApiRequestTracker(models.Model):
+    _name = 'mtd.api_request_tracker'
```
What does the HMRC API documentation call the thing you are tracking in this model?

ShreeOpusVL replied:
There is no name used for this we need to track this so we don't have another request for getting authorisation code.

ShreeOpusVL committed:
[IMP][O1758] change as per pr to remove the if block

and using getattr to call the functions.

Nick-OpusVL commented on account_mtd/models/mtd_hello_world.py:
```diff
@@ -0,0 +1,456 @@
+    def action_hello_world_connection(self):
+        if not self.hmrc_configuration:
+            raise exceptions.Warning("Please select HMRC configuration before continuing!")
+        # making sure that the endpoint is what we have used in the data file as the name can be changed anytime.
```
Why not store `mtd_hello_world_endpoint` in a field, and use that?

ShreeOpusVL replied:
using ID rather than the names. If they are changed in future then at least this would not cause errors. I have already had to change the name three times and had to remember to change the code accordingly. Which is why I have opted out for this option.

Nick-OpusVL replied:
So do they give us an id in some sort of response, then you pass it back to them later?

Nick-OpusVL commented on account_mtd/models/mtd_hello_world.py:
```diff
@@ -0,0 +1,456 @@
+        self.which_button_type_clicked = "helloworld"
+        self.path = "/hello/world"
+        version = self._json_command('version')
+        self._logger.info(
```
Logged this message in three places, and used it as a prefix in even more.  How about:

```
_logger.info(self._connection_button_clicked_log_message())
```

Where you've appended extra, you can include the result of that call in your format() calls or simply use `+` to concatenate your extra stuff.

Nick-OpusVL commented on account_mtd/models/mtd_hello_world.py:
```diff
@@ -0,0 +1,456 @@
+        self._logger.info(
+            "_json_command - has record_id been provided?:- {record_id}".format(record_id=record_id)
+        )
+        if record_id:
```
Any reason to accept in `record_id` rather than the record object itself?  Also what kind of record is it?  Call it what it is - `record` looks a bit too generic.

ShreeOpusVL replied:
I don't need the whole object as I will be doing a search on the id so that I can find the record and work with it. This is the hello world endpoint record_id, where the request of the connection had come from and I am already in the hello world class so it is called record _id

Nick-OpusVL replied:
Haven't you already got 'self' for that?

ShreeOpusVL replied:
No depending on what stage I may be at. i would not have self at that time. If I needed to get authorisation code then I would have to leave Odoo and go to HMRC webpage let the user follow the steps and redirect back to Odoo. therefore When we come back we do not have self. and we can't pass in self as parameter of redirect as HMRC does not return any other parameters than authorisation code.

Nick-OpusVL commented on account_mtd/models/mtd_hello_world.py:
```diff
@@ -0,0 +1,456 @@
+            self._logger.info(
+                "_json_command - we need to find the record and assign it to self"
+            )
+        # search the token table to see if we have access token
```
The code says this.  Comment not necessary.

Nick-OpusVL commented on account_mtd/models/mtd_hello_world.py:
```diff
@@ -0,0 +1,456 @@
+                "_json_command - we need to find the record and assign it to self"
+            )
+        # search the token table to see if we have access token
+        token_record = self.env['mtd.api_tokens'].search([('api_id', '=', self.api_name.id)])
```
`api_name.id` - this doesn't make sense.  Please explain.

Nick-OpusVL commented on account_mtd/models/mtd_hello_world.py:
```diff
@@ -0,0 +1,456 @@
+            )
+        # search the token table to see if we have access token
+        token_record = self.env['mtd.api_tokens'].search([('api_id', '=', self.api_name.id)])
+        access_token = token_record.access_token if token_record else ""
```
```
access_token = token_record.access_token or ""
```

If `token_record` is zero-length, then `token_record.access_token` will return `False`.

and

`(False or "") == ""`

Nick-OpusVL commented on account_mtd/models/mtd_hello_world.py:
```diff
@@ -0,0 +1,456 @@
+        # refresh_token = token_record.refresh_token if token_record else ""
+        
+        header_items = {"Accept": "application/vnd.hmrc.1.0+json"}
+        if self.which_button_type_clicked == "application":
```
This is smelling like strategy based on endpoint again, but not completely sure.

ShreeOpusVL replied:
Yes headers are different for different endpoints

Nick-OpusVL replied:
I didn't mean that, I was talking about the switching on which_button_type_clicked.

ShreeOpusVL replied:
Yes based on the endpoints, headers are different which is why I have used an internal variable to find out which endpoint I was in. Originally we just had one screen with three different buttons which is why I used the variable name "which_button_type_clicked."

Nick-OpusVL replied:
What does it mean now then?

ShreeOpusVL replied:
which endpoint did the user come from hello world, application or user. I will change the name of the variable.

Nick-OpusVL commented on account_mtd/models/mtd_hello_world.py:
```diff
@@ -0,0 +1,456 @@
+        )
+        response = requests.get(authorisation_url, timeout=3)
+        # response_token = json.loads(req.text)
+        self._logger.info(
```
I wonder if some of these belong at debug rather than info level.

Nick-OpusVL commented on account_mtd/models/mtd_hello_world.py:
```diff
@@ -0,0 +1,456 @@
+        else:
+            response_token = json.loads(response.text)
+            error_message = (
+                "Date {date}     Time {time} \n".format(date=datetime.utcnow().date(), time=datetime.utcnow().time())
```
Do you want to display times to users in UTC?

ShreeOpusVL replied:
yes in the time zone they are in right? That's what you suggested in the previous pr.

Nick-OpusVL replied:
No they're in Europe/London.  When BST starts these strings will be 1 hour out.

The ORM's date/datetime/time fields work in UTC (which is the context in which I suggested using utcnow()), but when formatting a string for humans you need to use local time.

Nick-OpusVL reviewed (changes requested):
Improved, but I have more questions and suggestions.  Some comments are in answer to Peter's so I hope it gives you the full conversation...

ShreeOpusVL committed:
[IMP][O1758] chanign variable name to new name

adding an extra selection option for the variable

ShreeOpusVL committed:
[IMP][O1758] Variable name change and format change

ShreeOpusVL committed:
[IMP][O1758] variable name pluralising

ShreeOpusVL committed:
[IMP][O1758] fields has to be relational to api model

ShreeOpusVL committed:
[IMP][O1758] Pep8 changes

ShreeOpusVL committed:
[IMP][O1758]Changes as per pr

ShreeOpusVL committed:
[IMP][O1758] removing a comment

ShreeOpusVL committed:
[IMp][O1758] attempting to reduce the function size

ShreeOpusVL committed:
[IMP][O1758] variable name change

adding an if to make sure we have a value for the error variable only
then we need to add a new line

ShreeOpusVL committed:
[IMP][O1758]Variable name change in xml and adding string name

ShreeOpusVL committed:
[IMP][O1758] making the function smaller

ShreeOpusVL committed:
[IMP][O1758] variable name change and removing utc.now

from date and time field

ShreeOpusVL committed:
[IMP][O1758]Variable name change

Needed to change the varible name in the data file as the variable name
in the module had been changed.

ShreeOpusVL committed:
[FIX][O1758]Request tracker needs to be closed

Whether the response is positive or negative we still need to close the
request tracker so that user can place a new request if required.

ShreeOpusVL committed:
[IMP][O1758] separating hello world and authorisation steps

We need to be able to inherit the hello world module for other mtd
modules and use the authorisation steps.
